README.md
bitcoin-python3 is a set of Python libraries that allows easy access to the
bitcoin peer-to-peer cryptocurrency client API. For Python 3 only.
To get Python 2 version - go to https://github.com/laanwj/bitcoin-python

Documentation
===========================

Documentation can be found here, or in the source archive. It is built
using Sphinx:

http://laanwj.github.com/bitcoin-python/doc/

Installation instructions
===========================

bitcoin-python3 uses setuptools for the install script. There are no dependencies apart from Python itself.

::

  $ python setup.py build
  $ python setup.py install

Pypi / Cheeseshop
==================

It is possible to install the package through Pypi (cheeseshop), see http://pypi.python.org/pypi?:action=display&name=bitcoin-python3

::

  $ pip install bitcoin-python3

TODO
======
These things still have to be added:

- SSL support (including certificate verification) for managing remote bitcoin daemons.



Changelog
2013-11-28  Ashot Seropian <ondaemon@gmail.com>

	* Python 3.3 support

2013-03-30  Wladimir van der Laan <laanwj@gmail.com>

	* Release 0.3

2013-03-30 16:13:06 +0100 Aviad Reich 

	* add keypoolrefill method (HEAD, origin/master, master)

2013-03-20 00:33:59 +0200 Aviad Reich 

	* raw transaction support added (rawtransactions)

2013-03-21 14:41:33 +0200 Aviad Reich 

	* cherry pick to fix dont_raise so it only blocks the correct exceptions to return a boolean (origin/wallet_locking)

2013-03-29 10:15:09 +0300 Aviad Reich 

	* ignore dist and python_egg

2013-03-23 19:08:46 +0000 Aviad Reich 

	* deprecated getblocknumber - use getblockcount instead

2013-03-19 23:01:18 +0200 Aviad Reich 

	* add wallet locking/unlocking support

2013-03-19 14:42:26 +0200 Aviad Reich 

	* in listtransactions, iterate all accounts if account is None (origin/listtransactions)

2013-03-19 11:16:38 +0200 Aviad Reich 

	* listaccounts can return a dictionary from account names to their balance if the param as_dict is True, update tests. (origin/listaccounts)

2013-03-19 09:44:20 +0200 Aviad Reich 

	* add tests for listunspent and getmininginfo

2013-03-18 08:18:23 +0200 Aviad Reich 

	* refactor __repr__

2013-03-17 09:43:44 +0200 Aviad Reich 

	* flake8 style conventions fixes

2013-03-18 00:13:42 +0200 Aviad Reich 

	* add listunspent method

2013-03-17 23:46:44 +0200 Aviad Reich 

	* add support for getmininginfo

2013-03-17 16:37:09 +0200 Aviad Reich 

	* Issue #15: add from parameter to listtransactions

2013-03-15 22:14:10 +0100 David Barton 

	* Fixed usage.rst broken links

2013-03-14 09:26:23 +0200 Aviad Reich 

	* connect_to_local(): use default testnet port, when the config file has testnet=x

2013-02-10 17:35:14 +0100 Adrian 

	* Add getrawtransaction method

2013-01-16 03:28:11 +0100 Thomas Steen Rasmussen 

	* fix bug when rpc reply error is null in json

2012-12-12 00:14:34 +0100 Benoît HERVIER 

	* There isn't always an error key in resp dict

2012-12-12 00:10:41 +0100 Benoît HERVIER 

	* Allow use of https

2012-10-29 16:46:18 +0100 Harald Schilly 

	* assert that we are really in the testnet

2012-10-28 23:25:25 +0100 Harald Schilly 

	* optionally allow to pass in the path to the bitcoin.conf file if it is at a non-standard path

2012-10-28 21:52:48 +0100 Harald Schilly 

	* fixing tests

2012-10-23 02:57:40 +0200 Thomas Steen Rasmussen 

	* Add support for the 'verifymessage' RPC command

2012-10-05 18:45:50 +0200 Wladimir J. van der Laan 

	* Switch to jgarzik's jsonrpc implementation

2012-10-05 18:32:23 +0200 Wladimir J. van der Laan 

	* Use RPC error codes from bitcoin core source

2012-07-15 12:07:53 +0400 Alexander Petrovich 

	* Added listsinceblock, getblock, getblockhash

2011-09-04 15:17:44 +0200 Jure Vrscaj 

	* Added sendmany command.

2011-08-24 14:53:53 +0200 Wladimir J. van der Laan 

	* allow listtransactions() (default to '' account), add test for gettransaction

2012-10-23  Thomas Steen Rasmussen <thomas@gibfest.dk>

	* Add "verifymessage" command

2011-08-24  Wladimir van der Laan <laanwj@gmail.com>

	* Add metadata for PyPi packaging

	* Add Changelog

	* Rename package to 'bitcoinrpc' instead of 'bitcoin', makes more sense as this is a RPC
	  client not a bitcoin client implementation.

	* Add COPYING (MIT license)

	* Add "gettransaction" command (thanks chopachom)


tests/test.py
# coding=utf-8
'''
Test script
*WARNING* Don't run this on a production bitcoin server! *WARNING*
Only on the test network.
'''
import argparse
import sys
sys.path.append('../src')

import bitcoinrpc
from bitcoinrpc.exceptions import BitcoinException, InsufficientFunds


from decimal import Decimal

parser = argparse.ArgumentParser()
parser.add_argument('--config', help="Specify configuration file")
parser.add_argument('--nolocal', help="Don't use connect_to_local",
                    action='store_true')
parser.add_argument('--noremote', help="Don't use connect_to_remote",
                    action='store_true')
args = parser.parse_args()

if __name__ == "__main__":

    if args.config:
        from bitcoinrpc.config import read_config_file
        cfg = read_config_file(args.config)
    else:
        from bitcoinrpc.config import read_default_config
        cfg = read_default_config(None)
    port = int(cfg.get('rpcport', '18332' if cfg.get('testnet') else '8332'))
    rpcuser = cfg.get('rpcuser', '')

    connections = []
    if not args.nolocal:
        local_conn = bitcoinrpc.connect_to_local()  # will use read_default_config
        connections.append(local_conn)
    if not args.noremote:
        remote_conn = bitcoinrpc.connect_to_remote(
                user=rpcuser, password=cfg['rpcpassword'], host='localhost',
                port=port, use_https=False)
        connections.append(remote_conn)

    for conn in connections:
        assert(conn.getinfo().testnet) # don't test on prodnet

        assert(type(conn.getblockcount()) is int)
        assert(type(conn.getconnectioncount()) is int)
        assert(type(conn.getdifficulty()) is Decimal)
        assert(type(conn.getgenerate()) is bool)
        conn.setgenerate(True)
        conn.setgenerate(True, 2)
        conn.setgenerate(False)
        assert(type(conn.gethashespersec()) is int)
        account = "testaccount"
        account2 = "testaccount2"
        bitcoinaddress = conn.getnewaddress(account)
        conn.setaccount(bitcoinaddress, account)
        address = conn.getaccountaddress(account)
        address2 = conn.getaccountaddress(account2)
        assert(conn.getaccount(address) == account)
        addresses = conn.getaddressesbyaccount(account)
        assert(address in addresses)
        #conn.sendtoaddress(bitcoinaddress, amount, comment=None, comment_to=None)
        conn.getreceivedbyaddress(bitcoinaddress)
        conn.getreceivedbyaccount(account)
        conn.listreceivedbyaddress()
        conn.listreceivedbyaccount()
        #conn.backupwallet(destination)
        x = conn.validateaddress(address)
        assert(x.isvalid == True)
        x = conn.validateaddress("invalid")
        assert(x.isvalid == False)
        messages = ('Hello, world!', 'かたな')
        for message in messages:
            signature = conn.signmessage(bitcoinaddress, message)
            assert(conn.verifymessage(bitcoinaddress, signature, message) is True)

        for accid in conn.listaccounts(as_dict=True).keys():
          tx = conn.listtransactions(accid)
          if len(tx) > 0:
            txid = tx[0].txid
            txdata = conn.gettransaction(txid)
            assert(txdata.txid == tx[0].txid)

        assert(type(conn.listunspent()) is list)  # needs better testing

        try:
            conn.sendtoaddress(bitcoinaddress, 10000000)
            assert(0)  # Should raise InsufficientFunds
        except InsufficientFunds:
            pass

    info = conn.getinfo()
    print("Blocks: %i" % info.blocks)
    print("Connections: %i" % info.connections)
    print("Difficulty: %f" % info.difficulty)

    m_info = conn.getmininginfo()
    print(("Pooled Transactions: {pooledtx}\n"
           "Testnet: {testnet}\n"
           "Hash Rate: {hashes} H/s".format(pooledtx=m_info.pooledtx,
                                            testnet=m_info.testnet,
                                            hashes=m_info.hashespersec)))


src/bitcoinrpc/proxy.py
"""
  Copyright (c) 2007 Jan-Klaas Kollhof
  Copyright (c) 2011-2013 Jeff Garzik
  Copyright (c) 2013 Nikolay Belikov (nikolay@belikov.me)
  Copyright (c) 2013 Ashot Seropian (ondaemon@gmail.com)


  jsonrpc is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation; either version 2.1 of the License, or
  (at your option) any later version.

  This software is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this software; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
"""

import base64
import json
import decimal
from collections import defaultdict, deque
from bitcoinrpc.exceptions import TransportException
import urllib
import http

USER_AGENT = "AuthServiceProxy/0.1"

HTTP_TIMEOUT = 30


class JSONRPCException(Exception):
    def __init__(self, rpc_error):
        Exception.__init__(self)
        self.error = rpc_error


class HTTPTransport(object):
    def __init__(self, service_url):
        self.service_url = service_url
        self.parsed_url = urllib.parse.urlparse(service_url)
        if self.parsed_url.port is None:
            port = 80
        else:
            port = self.parsed_url.port
        authpair = "%s:%s" % (self.parsed_url.username,
                              self.parsed_url.password)
        authpair = authpair.encode('utf8')
        self.auth_header = "Basic ".encode('utf8') + base64.b64encode(authpair)
        if self.parsed_url.scheme == 'https':
            self.connection = http.client.HTTPSConnection(self.parsed_url.hostname,
                                                      port, None, None, False,
                                                      HTTP_TIMEOUT)
        else:
            self.connection = http.client.HTTPConnection(self.parsed_url.hostname,
                                                     port, False, HTTP_TIMEOUT)

    def request(self, serialized_data):
        self.connection.request('POST', self.parsed_url.path, serialized_data,
                                {'Host': self.parsed_url.hostname,
                                 'User-Agent': USER_AGENT,
                                 'Authorization': self.auth_header,
                                 'Content-type': 'application/json'})

        httpresp = self.connection.getresponse()
        if httpresp is None:
            self._raise_exception({
                'code': -342, 'message': 'missing HTTP response from server'})
        elif httpresp.status == http.client.FORBIDDEN:
            msg = "bitcoind returns 403 Forbidden. Is your IP allowed?"
            raise TransportException(msg, code=403,
                                     protocol=self.parsed_url.scheme,
                                     raw_detail=httpresp)

        resp = httpresp.read()
        return resp.decode('utf8')


class FakeTransport(object):
    """A simple testing facility."""
    def __init__(self):
        self._data = defaultdict(deque)

    def load_serialized(self, method_name, fixture):
        self._data[method_name].append(fixture)

    def load_raw(self, method_name, fixture):
        self._data[method_name].append(json.dumps(fixture))

    def request(self, serialized_data):
        data = json.loads(serialized_data, parse_float=decimal.Decimal)
        method_name = data['method']
        return self._data[method_name].popleft()


class RPCMethod(object):
    def __init__(self, name, service_proxy):
        self._method_name = name
        self._service_proxy = service_proxy

    def __getattr__(self, name):
        new_name = '{}.{}'.format(self._method_name, name)
        return RPCMethod(new_name, self._service_proxy)

    def __call__(self, *args):
        self._service_proxy._id_counter += 1
        data = {'version': '1.1',
                'method': self._method_name,
                'params': args,
                'id': self._service_proxy._id_counter}
        postdata = json.dumps(data)
        resp = self._service_proxy._transport.request(postdata)
        resp = json.loads(resp, parse_float=decimal.Decimal)

        if resp['error'] is not None:
            self._service_proxy._raise_exception(resp['error'])
        elif 'result' not in resp:
            self._service_proxy._raise_exception({
                'code': -343, 'message': 'missing JSON-RPC result'})
        else:
            return resp['result']

    def __repr__(self):
        return '<RPCMethod object "{name}">'.format(name=self._method_name)


class AuthServiceProxy(object):
    """
    You can use custom transport to test your app's behavior without calling
    the remote service.

    exception_wrapper is a callable accepting a dictionary containing error
    code and message and returning a suitable exception object.
    """
    def __init__(self, service_url, transport=None, exception_wrapper=None):
        self._service_url = service_url
        self._id_counter = 0
        self._transport = (HTTPTransport(service_url) if transport is None
                           else transport)
        self._exception_wrapper = exception_wrapper

    def __getattr__(self, name):
        return RPCMethod(name, self)

    def _get_method(self, name):
        """
        Get method instance when the name contains forbidden characters or
        already taken by internal attribute.
        """
        return RPCMethod(name, self)

    def _raise_exception(self, error):
        if self._exception_wrapper is None:
            raise JSONRPCException(error)
        else:
            raise self._exception_wrapper(error)
